### PR TITLE
Laboratory work #3, Tasya Kakoulina - 23FPL2.

### PR DESCRIPTION
**Pull request by tasya-kakoulina**

### CODE REVIEW

File: lab_3_ann_retriever/main.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 # pylint: disable=too-few-public-methods, too-many-arguments, duplicate-code, unused-argument
 from typing import Protocol
-
+import math
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:9]: new line

File: lab_3_ann_retriever/start.py [hunk 2]
```diff
@@ -28,9 +29,20 @@ def main() -> None:
     """
     Launch an implementation.
     """
-    with open("assets/secrets/secret_1.txt", "r", encoding="utf-8") as text_file:
+    with open("assets/secrets/secret_2.txt", "r", encoding="utf-8") as text_file:
         text = text_file.read()
-    result = None
+    secret_vector = tuple(float(item) for item in text.split(', '))
+    result = ':('
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/start.py:35]: amazing move

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -541,6 +638,19 @@ def retrieve_relevant_documents(
...
+        if (not query or not isinstance(query, str) or not isinstance(n_neighbours, int)
+                or isinstance(n_neighbours, bool)):
+            return None
+        query_vector = self._index_document(query)
+        if not query_vector:
+            return None
+        knn_list = self._calculate_knn(query_vector, self._document_vectors, n_neighbours)
+        if not knn_list:
+            return None
+        result = []
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:650]: list comprehension

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_ann_retriever/main.py (hunk 1 above)
Comment (tasya-kakoulina) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -194,6 +254,13 @@ def save(self, file_path: str) -> bool:
         Returns:
             bool: True if saved successfully, False in other case
         """
+        if not file_path or not isinstance(file_path, str):
+            return False
+        with open(file_path, 'w', encoding='utf-8') as file:
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:259]: json files contains unicode - try to save with russian letters

File: lab_3_ann_retriever/start.py (hunk 2 above)
Comment (artyomtugaryov) [reply lab_3_ann_retriever/start.py:37]: So, remove it now

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -183,6 +234,15 @@ def vector2tokens(self, vector: Vector) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not vector or len(vector) != len(self._token2ind):
+            return None
+        result = []
+        for i, num in enumerate(vector):
+            if num != 0.0:
```
[the remark this answers is not in the document]
Comment (tasya-kakoulina) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py (hunk 4 above)
Comment (tasya-kakoulina) [reply lab_3_ann_retriever/main.py:265]: done

File: lab_3_ann_retriever/start.py (hunk 2 above)
Comment (tasya-kakoulina) [reply lab_3_ann_retriever/start.py]: done

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -49,6 +52,15 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
...
     """
+    if query_vector is None or document_vector is None:
+        return None
+    if not query_vector or not document_vector:
+        return 0.0
+    sum_ = 0.0
+    for ind, value in enumerate(query_vector):
+        sum_ += (value - document_vector[ind]) ** 2
+    sum_ = sum_ ** 0.5
+    return sum_ if isinstance(sum_, float) else None
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:63]: If not float, what else?

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_3_ann_retriever/main.py:872]: where?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good!

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (tasya-kakoulina) [reply lab_3_ann_retriever/main.py]: sorry, i misunderstood it.. is it better now?

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -49,6 +52,15 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
...
     In case of corrupt input arguments, None is returned.
     """
+    if query_vector is None or document_vector is None:
+        return None
+    if not query_vector or not document_vector:
+        return 0.0
+    sum_ = 0.0
+    for ind, value in enumerate(query_vector):
+        sum_ += (value - document_vector[ind]) ** 2
+    sum_ = sum_ ** 0.5
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:62]: return as is, without setting to the variable

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Looks good to me. Only small comment

**Pull request closed without merging**